### PR TITLE
feat: check Units for Loadflow variables

### LINKED ISSUE #23: feat: check Units for Loadflow variables
** Problem**
* The exponents of variable units (like kW or MW) are set in the project settings by default. The exporter assume, that default exponents are set, but the user could have adjusted them.
* For LV or MV loads the unit exponent is set as element variable setting (not influenced by global setting) and therefore has to be checked seperatly.

** Solution**
* save project settings
* set default unit exponents
* export
* restore project settings

### PR DESCRIPTION
**Pull request by sasanjac**
feat: check Units for Loadflow variables
fixes #23

### CODE REVIEW

**Comment by SebastianDD** [comment]
consider for test case .ElmPvsys: because there active power (pgini_a) is given in MW and reactive power (qgini_a) is given in kvar

**Comment by sasanjac** [comment]
1. all settings regarding units are saved at the start of the program. 
2. all unit prefixes get set to the defined ones (power: `M`, length: `k`, voltage: `k`, current: `k`).
3. all custom unit conversions are deleted.
4. the following unit conversions are applied: `PVSystem`, `LVLoad`. Here we will have to add/adjust in the future as we find other prefixes deviating from the defined ones.
5. at the end, all unit conversion settings are reset to their original states. To ensure that this is actually applied even in the case of an error, I have added a context manager for the `exporter` which will take care of that.

File: powerfactory_utils/interface.py [hunk 1]
```diff
@@ -36,19 +60,39 @@ def __post_init__(self) -> None:
...
 
-        self.grid_data = self.load_grid_data()
-        self.study_case_lib = self.load_study_case_lib()
-        self.scenario_lib = self.load_scenario_lib()
-        self.grid_graphs_dir = self.load_grid_graphs()
+        self.ext_data_dir = self.project_settings_dir.extDataDir
+
+        self.grid_data = self.app.GetProjectFolder("netdat")
+        self.study_case_dir = self.app.GetProjectFolder("study")
+        self.scenario_dir = self.app.GetProjectFolder("dia")
```
Comment (PietzoJo) [inline powerfactory_utils/interface.py:84]:
```suggestion
        self.scenario_dir = self.app.GetProjectFolder("scen")
```

File: powerfactory_utils/constants.py [hunk 2]
```diff
@@ -16,3 +21,23 @@ class DecimalDigits:
...
     POWER = 0
     PU = 4
+
+
+class BaseUnits:
+    LENGTH: pft.MetricPrefix = "k"
+    POWER: pft.MetricPrefix = "M"
+    CURRENCY: pft.Currency = "EUR"
+    UNITCONVERSIONS: dict[str, list[UnitConversion]] = {
+        "Elmlodlv": [
```
Comment (PietzoJo) [inline powerfactory_utils/constants.py:31]:
```suggestion
        "ElmLodlv": [
```

**Review by PietzoJo — DISMISSED** [review]

**Comment by PietzoJo** [comment]
> 1. all settings regarding units are saved at the start of the program.
> 
>     2. all unit prefixes get set to the defined ones (power: `M`, length: `k`, voltage: `k`, current: `k`).
> 
>     3. all custom unit conversions are deleted.
> 
>     4. the following unit conversions are applied: `PVSystem`, `LVLoad`. Here we will have to add/adjust in the future as we find other prefixes deviating from the defined ones.
> 
>     5. at the end, all unit conversion settings are reset to their original states. To ensure that this is actually applied even in the case of an error, I have added a context manager for the `exporter` which will take care of that.

Since all prefixes are set to the defined once with set_default_unit_conversion(), are these correction factors in export.py still accurate?

exporter.py-->line 63:
`LV_TO_BASE_POW = Exponents.LV_POWER / Exponents.POWER
LV_TO_BASE_CURR = Exponents.LV_CURRENT / Exponents.CURRENT`

exporter.py-->line 525:
`i_r = line.nlnum * line.fline * i * Exponents.CURRENT  # rated current (A)`

exporter.py-->line 882:
`def calc_load_lv_power(self, load: pft.LoadLV) -> LoadLV:
        load_type = load.iopt_inp
        scaling = load.scale0
        if not load.i_sym:
            power_fixed = self.calc_load_lv_power_fixed_sym(load, scaling)
        else:
            if load_type == 0:
                power_fixed = LoadPower.from_sc_asym(
                    s_r=load.slinir * LV_TO_BASE_POW,
                    s_s=load.slinis * LV_TO_BASE_POW,
                    s_t=load.slinit * LV_TO_BASE_POW,
                    cosphi_r=load.coslinir,
                    cosphi_s=load.coslinis,
                    cosphi_t=load.coslinit,
                    scaling=scaling,
                )
            elif load_type == 1:
                power_fixed = LoadPower.from_pc_asym(
                    p_r=load.plinir * LV_TO_BASE_POW,
                    p_s=load.plinis * LV_TO_BASE_POW,
                    p_t=load.plinit * LV_TO_BASE_POW,
                    cosphi_r=load.coslinir,
                    cosphi_s=load.coslinis,
                    cosphi_t=load.coslinit,
                    scaling=scaling,
                )
            elif load_type == 2:
                power_fixed = LoadPower.from_ic_asym(
                    u=load.ulini,
                    i_r=load.ilinir * LV_TO_BASE_CURR,
                    i_s=load.ilinis * LV_TO_BASE_CURR,
                    i_t=load.ilinit * LV_TO_BASE_CURR,`
                    
I did not find out how to comment it directly in the code, so I printed it here.

**Comment by SebastianDD** [comment]
it seems further, that the created the new default unit conversion settings (cfg. line 67) are kind of not applied to the grid model.
The same behavior can be observed when creating `.SetVariable` by hand in PowerFactory. One have to press "apply". --> PowerFactory Support?

**Comment by SebastianDD** [comment]
> it seems further, that the created the new default unit conversion settings (cfg. line 67) are kind of not applied to the grid model. The same behavior can be observed when creating `.SetVariable` by hand in PowerFactory. One have to press "apply". --> PowerFactory Support?

I'm going to ask the PowerFactory Support for the best solution. An alternative would be the deactivation and activation of the PowerFactory project after unit conversion setup resp. before the export to apply the unit conversion changes.

**Review by SebastianDD — APPROVED** [review]
had to fix 2 issues and adapt example code

**Review by SebastianDD — APPROVED** [review]

**Pull request merged**